text/0000-infered-types.md

```diff
@@ -0,0 +1,149 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When crating a struct or enum, infered types can simplify the type into just a underscore. It is important to note, however, that they do not work when the type is not specific enough to be infered like: type parameters. Below are some examples where they do and don't work.
+
+Function calls (structs):
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+// my_function(MyStruct {
+//     value: 1
+// });
+my_function(_ {
+   value: 1
+});
+```
+
+Function calls (impl):
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+// my_function(MyStruct::new()});
+my_function(_::new());
+```
+
+Function returns (enum):
+```rust
+fn my_function() -> MyEnum {
+   // MyEnum::MyVarient
+   _::MyVarient
+}
+```
+
+Match arms:
+```rust
+enum Example {
+   One,
+   Two
+}
+
+fn my_fn(my_enum: Example) -> String {
+   match my_enum {
+      _::One => "One!",
+      _::Two => "Two!"
+   }
+}
+```
+
+It is important to note that `_` only represents the type; if you have (for example) another enum that can be coerced into the type, you will need to specify it manually. Additionally, any traits required to call an impl will still have to be imported.
+
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+
+my_function(MyStruct2::do_something().into()); // ✅
+
+
+my_function(_::do_something().into()); // ❌ error[E0599]: variant or associated item not found in `MyStruct`
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The underscore operator infers the type of a stuct or enum where there is enough type information in a context in order to infer an exact type. In a statement like `_::Variant`, you can imagine the underscore to be the type. That means that anything you could do with an actual type like `MyEnum::Variant` will still apply in an infered enum. Ultimately, the enum or struct doesn't need to be imported but, traits and other specfied things will need to be imported.
+
+Due to how the rust compiler currently works, lots of changes will need to be made to allow paths to be infered in an order that allows for all of the mentioned. One issue is getting the type of a given context mentioned in [rust-lang/rust#8995](https://github.com/rust-lang/rust/issues/8995).
+
+Finally, here are some examples of non-strict typings that can not be allowed.
+```rust
+fn convert<T>(argument: T) -> Example {/* ... */}
+
+do_something(convert(_::new()))
+//                   ^^^^^^^^ Cannot infer type on generic type argument
+```
+
+However, ones where a generic argument can collapse into strict typing can be allowed. The below works because `T` becomes `Example`. This wouldn’t work if `do_something`, however, took a generic.
+```rust
+fn do_something(argument: Example) {/* ... */}
+fn convert<T>(argument: T) -> T {/* ... */}
+
+do_something(convert(_::new()))
+```
```

Comment:
I am replacing this entire section since examples are not "reference". Consider using the proposed specification of semantics:

```suggestion
In particular, when an "underscore-type expression" appears as the root expression at an "inferable expression context" and the type expected at the context can be exactly inferred without additional information, the `_` is desugared into the expected type at the context.

"Underscore-type expression" is defined as one of the following:

- A single underscore. This is expected to be a unit struct constructor.
- A named struct literal with the type as `_` (e.g. `_ { field1, field2 }`)
- A single-segment function call expression with the path `_` (e.g. `_(field1, field2)`). This is expected to be a tuple struct constructor.
- A two-segment path expression with the first segment as `_` (e.g. `_::EnumVariant`, `_::CONSTANT`)
- A two-segment function call expression with the first segment as `_` (e.g. `_::fn_call()`)
- An underscore-type expression as the subject of a method call (e.g. `underscore_expr.method()`)
- An underscore-type expression in a try expression (e.g. `underscore_expr?`)
- An underscore-type expression in an await expression (e.g. `underscore_expr.await`)

An "inferable expression context" is defined as one of the following:

- A value in a function/method call argument list
- The value of a field in a struct literal
- The value of a value in an array literal
- An operand in a range literal (i.e. if an expression is known to be of type `Range<T>`, `expr..expr` can infer that both `expr`s are of type `T`)
- The value after `return`/`break`/`yield`
```

Notable **non**-inferable expression contexts include:

- Binary operators (because it is very frequent that binary operators do not preserve the parent type)

Similarly, when an "underscore-type pattern" appears as the root pattern at an "inferable pattern context", where "underscore-type pattern" is defined as one of the following:

- The underscore as the only segment of a single-segment-path named or tuple struct (e.g. `_ { field1 }`, `_(field1)`
- The underscore as the first segment of a two-segment-path named or tuple struct (e.g. `_::Variant { field1 }`, `_::Variant(field1)`)
- The underscore as the first segment of a two-segment path (e.g. `_::Variant`)

A single `_` is not an underscore-type pattern because it is already the wildcard pattern.

Reply:
Feels like a legal document 😵‍💫